runtime/json-functions.cpp

```diff
@@ -215,12 +215,41 @@ bool do_json_encode_string_vkext(const char *s, int len) noexcept {
 
 namespace impl_ {
 
+string JsonPath::to_string() const {
+  // this function is called only when error is occurred, so it's not
+  // very performance-sensitive
+
+  if (depth == 0) {
+    return string{"/", 1};
+  }
+  unsigned num_parts = std::clamp(depth, 0U, static_cast<unsigned>(arr.size()));
+  string result;
+  result.reserve_at_least((num_parts+1) * 8);
+  result.push_back('/');
+  for (unsigned i = 0; i < num_parts; i++) {
+    const char *key = arr[i];
+    if (key == nullptr) {
+      // int key indexing
+      result.append("[.]");
+    } else {
+      // string key indexing
+      result.append("['");
+      result.append(arr[i]);
+      result.append("']");
+    }
+  }
+  if (depth >= arr.size()) {
+    result.append("...");
+  }
+  return result;
+}
+
 JsonEncoder::JsonEncoder(int64_t options, bool simple_encode) noexcept:
   options_(options),
   simple_encode_(simple_encode) {
 }
 
-bool JsonEncoder::encode(bool b) const noexcept {
+bool JsonEncoder::encode(bool b) noexcept {
```

Comment:
why did you remove `const` here and in some places below?

Reply:
Because now encoder methods modify the internal state - json path object.
It's more efficient than passing a copy to every next frame.

Reply:
Yes, but I don't see, where it modifies `json_path_` in this method or in `encode(int64_t i)` for example

Reply:
I would say that it looks like an oversight. I'll return const qualifiers to the places where that's possible.

I don't remember if there was a real reason to remove them in these places.

Reply:
I remembered now.
Having different constness there makes implicit T->Optional conversion a viable candidate for cases like `encode(true)` which we don't want.
Since we're dispatching the encoding like `encode(value)`, it's cumbersome to rename functions to `encode_int` and so on to disambiguate between overloadings. Removing const qualifiers from encode<int> and encode<bool> looked like a lesser evil.